### PR TITLE
Multiclient Update 

### PR DESCRIPTION
**Pull request by Chenyu-Shi**
Added User Class in core.py
Added public keys storage in enclave and helpful function that verifies any user's signature 
Hardcoded preapproved user list

### CODE REVIEW

File: python-package/securexgboost/core.py [hunk 1]
```diff
@@ -355,7 +355,74 @@ def _maybe_dt_array(array):
...
+        _check_call(_LIB.sign_data(keyfile, data, data_size, signature, ctypes.byref(sig_len)))
+
+        # Cast the signature and sig_len back to a gRPC serializable format
+        sig_len_as_int = sig_len.value
+        signature = pointer_to_proto(signature, sig_len_as_int)
+
+        return signature, sig_len_as_int
+
+
+def Set_user(user_name):
```
Comment (chester-leung) [inline python-package/securexgboost/core.py]: You've moved this function into the User class right? Can we delete it?

File: python-package/securexgboost/__init__.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 
 import os
 
-from .core import DMatrix, Booster, Enclave, CryptoUtils, set_user
+from .core import DMatrix, Booster, Enclave, CryptoUtils, Set_user, User
```
Comment (chester-leung) [inline python-package/securexgboost/__init__.py]: I think we can remove `Set_user` here if it's in the User class.

File: python-package/securexgboost/__init__.py [hunk 3]
```diff
@@ -23,7 +23,7 @@
     __version__ = f.read().strip()
 
 __all__ = ['DMatrix', 'Booster', 'Enclave', 'CryptoUtils', 
-           'train', 'cv', 'set_user', 
+           'train', 'cv', 'Set_user', 'User',
```
Comment (chester-leung) [inline python-package/securexgboost/__init__.py]: Here too.

File: python-package/securexgboost/core.py [hunk 4]
```diff
@@ -2124,4 +2191,4 @@ def get_split_value_histogram(self, key, feature, fmap='', bins=None, as_pandas=
         if as_pandas and not PANDAS_INSTALLED:
             sys.stderr.write(
                 "Returning histogram as ndarray (as_pandas == True, but pandas is not installed).")
-        return nph
+        return nph
```
Comment (chester-leung) [inline python-package/securexgboost/core.py]: Please revert this

File: enclave/include/enclave_context.h [hunk 5]
```diff
@@ -319,13 +365,18 @@ class EnclaveContext {
...
-      std::vector<uint8_t> v(output, output + CIPHER_KEY_SIZE);
+      // storing user private key
+      std::vector<uint8_t> user_private_key(output, output + CIPHER_KEY_SIZE);
       std::string user_nam =  convertToString((char *)nameptr, name_len);
-      client_keys.insert({user_nam, v});
+      client_keys.insert({user_nam, user_private_key});
+
+      // storing user public key
+      // TODO verify that user certificate's public key has the same length as the secure enclave public key
+      mbedtls_rsa_alt_context * rsa_info = (mbedtls_rsa_alt_context *) &(user_cert.pk.pk_info);
```
Comment (chester-leung) [inline enclave/include/enclave_context.h]: To stay consistent with formatting, let's either put the `*` together with the type or the variable name, i.e. `mbedtls_rsa_alt_context* rsa_info` or `mbedtls_rsa_alt_context *rsa_info`.

Comment (Chenyu-Shi) [reply enclave/include/enclave_context.h]: fixed

File: python-package/securexgboost/core.py (hunk 4 above)
Comment (Chenyu-Shi) [reply python-package/securexgboost/core.py]: reverted i think? Added a newline at the end

File: python-package/securexgboost/__init__.py (hunk 2 above)
Comment (Chenyu-Shi) [reply python-package/securexgboost/__init__.py]: deleted

File: python-package/securexgboost/core.py (hunk 1 above)
Comment (Chenyu-Shi) [reply python-package/securexgboost/core.py]: yep deleted

**Pull request merged**